src/ui/public/visualize/visualization.js

```diff
@@ -0,0 +1,147 @@
+    link: function ($scope, $el) {
+      const minVisChartHeight = 180;
+      const resizeChecker = new ResizeChecker($el);
+
+      $scope.showSpyPanel = $scope.vis && $scope.vis.showSpyPanel || false;
+
+
+      //todo: lets make this a simple function call.
+      const getVisEl = jQueryGetter('.visualize-chart');
+      const getVisContainer = jQueryGetter('.vis-container');
+      const getSpyContainer = jQueryGetter('.visualize-spy-container');
+
+      // Show no results message when isZeroHits is true and it requires search
+      $scope.showNoResultsMessage = function () {
+        const requiresSearch = _.get($scope, 'vis.type.requiresSearch');
+        const isZeroHits = _.get($scope,'visData.hits.total') === 0;
+        const shouldShowMessage = !_.get($scope, 'vis.params.handleNoResults');
+
+        return Boolean(requiresSearch && isZeroHits && shouldShowMessage);
+      };
+
+      const legendPositionToVisContainerClassMap = {
+        top: 'vis-container--legend-top',
+        bottom: 'vis-container--legend-bottom',
+        left: 'vis-container--legend-left',
+        right: 'vis-container--legend-right',
+      };
+
+      $scope.getVisContainerClasses = function () {
+        return legendPositionToVisContainerClassMap[$scope.vis.params.legendPosition];
+      };
+
+      $scope.spy = {};
+      $scope.spy.mode = ($scope.uiState) ? $scope.uiState.get('spy.mode', {}) : {};
+
+      const applyClassNames = function () {
+        const $visEl = getVisContainer();
+        const $spyEl = getSpyContainer();
+        if (!$spyEl) return;
+
+        const fullSpy = ($scope.spy.mode && ($scope.spy.mode.fill || $scope.fullScreenSpy));
+
+        $visEl.toggleClass('spy-only', Boolean(fullSpy));
+        $spyEl.toggleClass('only', Boolean(fullSpy));
+
+        $timeout(function () {
+          if (shouldHaveFullSpy()) {
+            $visEl.addClass('spy-only');
+            $spyEl.addClass('only');
+          }
+        }, 0);
+      };
+
+      const loadingDelay = config.get('visualization:loadingDelay');
+      $scope.loadingStyle = {
+        '-webkit-transition-delay': loadingDelay,
+        'transition-delay': loadingDelay
+      };
+
+      function shouldHaveFullSpy() {
+        const $visEl = getVisEl();
+        if (!$visEl) return;
+
+        return ($visEl.height() < minVisChartHeight)
+          && _.get($scope.spy, 'mode.fill')
+          && _.get($scope.spy, 'mode.name');
+      }
+
+      // spy watchers
+      $scope.$watch('fullScreenSpy', applyClassNames);
+
+      $scope.$watchCollection('spy.mode', function () {
+        $scope.fullScreenSpy = shouldHaveFullSpy();
+        applyClassNames();
+      });
+
+      const Visualization = $scope.vis.type.visualization;
+
+      //todo: make this not a jquery element
+      const visualization = new Visualization(getVisEl()[0], $scope.vis);
+
+      const renderFunction = _.debounce(() => {
+        visualization.render($scope.visData)
```

Comment:
by calling `Visualization#render()` with the data for any changes (even if the data has not changed or not), it makes it cumbersome for visualizations to determine if the render-function is called due to changes in the data, changes in the params, or both. This is important for a visualizations with persistent UI that animate and make immediate changes.

Also, because of the assymetry of how we handle `params` and `esReponse`. Former is proeprty of the vis-object, latter is argument into `render`.
- Any thoughts on adding methods/state to the vis-object, or extra parameters to the `render` method, to help users determine why exactly render is called? 
- Alternatively, we could pass in two-parameters into render. The `esReponse` and the `params`, and remove params from the vis-object.